### ISSUE TITLE
Synchronous IO

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#1503]
Tons of existing APIs for IO are synchronous. For instance, the http.server API #140. Also, basic things like `input()` and any tool that expects to do CLI interaction are based around synchronous IO.

@joemarshall has implemented a really clever solution to this here: https://github.com/joemarshall/

It would be a huge improvement to incorporate this approach.

**Comment by jprendes** [comment]
This sounds like a perfect use case for [Asyncify](https://emscripten.org/docs/porting/asyncify.html).
Looks like @mfripp was in the right direction [here](https://github.com/pyodide/pyodide/issues/1219#issuecomment-776361057) but didn't have the bandwidth to take this on.

Is there any flag against using Asyncify? I guess the obvious ones are binary size and performance, although I would expect better performance than with unthrow.

**Comment by jprendes** [comment]
What do you mean by spooky?

I've put together a working example using Asyncify here: https://github.com/jprendes/pyodide

The built demo is here: https://jprendes.github.io/pyodide/

Try running the following code:
```python
from pyodide import await_js, to_js
from js import sleep, performance

for i in range(5):
    start = performance.now();
    await_js(to_js(sleep(1000)));
    end = performance.now();
    print(f"{i}: slept for {round(end-start)} ms");

```

**Comment by jprendes** [comment]
I haven't measured the performance with/without Asyncify, but from Alon's talk [here](https://youtu.be/qQOP6jqZqf8?t=1060) it looks like the expected overhead shouldn't be too much (around 50%).

**Comment by jprendes** [comment]
Thanks, I actually needed this functionality for a personal project in doing. 

> That I am worried it could have drawbacks but I'm not sure how to reason about those possible drawbacks.

I'd be happy to discuss about it with you. Other than binary size, and maybe performance, the other key aspect is that the shared libraries may need to be compiled with asincify too. But that would depend on the actual CPython code, which I'm not familiar with. 

> We could avoid the to_js by using a call structure like syncify_js_func(sleep)(1000).

That a great idea! I'll give it a go. 

> Maybe we could offer a call_asyncify entrypoint and one without it? Would the one without call_asyncify avoid the 50% performance cost?

I don't want to say it's impossible, but probably not the right approach either. It would not be supported by emscripten out of the box and require quite some hacking. Maybe two builds, with and without asincify would be better.

Also, the 50% figure I mentioned is based on the talk I linked, but I think it would be better to actually measure it with a benchmark. 

The performance overhead is not in those js code changes. The main overhead is on the wasm side, to allow the code to be paused and resumed. The JS changes there are just to make sure a promise is returned if asincify kicked in, and a normal value otherwise to preserve the normal behaviour.

**Comment by hoodmane** [comment]
> the shared libraries may need to be compiled with asincify too. But that would depend on the actual CPython code, which I'm not familiar with.

I suspect this won't be needed. I'd assume the only issue is if they try to call an asyncified callback?

> it would be better to actually measure it with a benchmark

Yes, please benchmark the performance!

> two builds, with and without asincify would be better

Sounds like it. I guess the alternative would be to basically double wasm code size and compilation time by having an asyncified and non asyncified copy of each function...

**Comment by joemarshall** [comment]
Good news is that it doesn't need cors headers to be set or anything. Bad news is that it doesn't work in incognito mode because that bans service workers on some browsers.

**Comment by stefnotch** [comment]
@joemarshall Oh nice, so you've implemented that option!

Relevant comment: https://github.com/pyodide/pyodide/issues/1504#issuecomment-882105354

**Comment by ryanking13** [comment]
@joemarshall That sounds very nice! Would you like to share us your work?

**Comment by hoodmane** [comment]
IIRC I think @alexmojaki may have also done something like this.

**Comment by winniederidder** [comment]
> IIRC I think @alexmojaki may have also done something like this.

The described input handling is indeed implemented in [sync-message](https://github.com/alexmojaki/sync-message).

**Comment by stefnotch** [comment]
> > IIRC I think @alexmojaki may have also done something like this.
> 
> The described input handling is indeed implemented in [sync-message](https://github.com/alexmojaki/sync-message).

Oh my, is that essentially a synchronous version of [comlink](https://github.com/GoogleChromeLabs/comlink)?
(Making it quite similar to https://github.com/hoodmane/synclink or the stuff in [starboard](https://github.com/gzuidhof/starboard-notebook/tree/master/packages/starboard-python/src/worker), except for the quite different implementation details)

**Comment by alexmojaki** [comment]
sync-message is low level and can be used with comlink or just vanilla `postMesage`.

[comsync](https://github.com/alexmojaki/comsync) (the name is inspired by synclink) combines comlink and sync-message to make things a bit easier and adds a bit more functionality.

[pyodide-worker-runner](https://github.com/alexmojaki/pyodide-worker-runner/) then combines comsync and [python_runner](https://github.com/alexmojaki/python_runner) to help with a variety of things.

These are still quite new, I was hoping for a bit more time before announcing them :sweat_smile:

They're mostly stable, although I expect to make a couple more changes to pyodide-worker-runner as new versions of Pyodide are released, and python_runner needs documentation.

I created these to share code between futurecoder and Papyros (by @winniederidder) and hopefully other Pyodide projects.

**Comment by hoodmane** [comment]
I made a post about this in the new `discuss.python.org`
https://discuss.python.org/t/wasm-python-synchronous-io-working-group/15509

Unfortunately we need Javascript devs more than Python devs...

**Comment by alexmojaki** [comment]
pyodide_worker_runner uses comsync, which uses both sync_message and comlink, the last of which is just a general worker communication thing unrelated to sync IO. All of these work and are fully in use in https://futurecoder.io/ and https://papyros.dodona.be/.

Unfortunately there are still many parts to digest and I think there's a limit to how much can be done about that. I recommend going bottom-up, starting by getting something basic working with the lower level libraries to understand them. Starting with https://github.com/alexmojaki/sync-message - you shouldn't have to use most of the functions there if you can use the higher-level libraries, but you will still have to use `makeChannel()`, and critically you will have to set up cross-origin isolation and/or a service worker and verify that it works.

After sync-message, get an understanding of comlink, then comsync. None of these know anything about Pyodide, so your experiments don't have to involve them, just make sure you can communicate in the way you want.

I'm working on expanding/updating the docs of pyodide_worker_runner in this branch: https://github.com/alexmojaki/pyodide-worker-runner/tree/docs. In addition to the README, you may find it helpful to look at the 'docstrings' in https://github.com/alexmojaki/pyodide-worker-runner/blob/docs/lib/index.ts. I also hope to eventually write one or more blog posts explaining all these libraries and how they fit together. Note that the library also includes some features unrelated to sync IO - maybe they can still be useful to you, but don't let them confuse you or distract from the sync IO goal. Finally, it might help to see real life code using it on the [main thread](https://github.com/alexmojaki/futurecoder/blob/master/frontend/src/TaskClient.js) and [worker](https://github.com/alexmojaki/futurecoder/blob/master/frontend/src/Worker.js).

**Comment by hoodmane** [comment]
@alexmojaki Do you have anything set up to do keyboard interrupts with the service worker / XMLHttpRequest approach? I think you could do this by calling `pyodide.setInterruptBuffer(proxy)` where `proxy` has a get handler that looks at the status of an XMLHttpRequest that the service worker resolves to interrupt execution. But I haven't tried it.

**Comment by alexmojaki** [comment]
No, I haven't tried anything like that. I'd be curious about the performance impact of regularly checking that throughout Python execution. The helpers that pyodide-worker-runner/comsync provide for interruption are:

1. Smoothly interrupting sync IO (reading stdin or sleeping) without the help of `setInterruptBuffer`.
2. Making `setInterruptBuffer` easier to use when possible.
3. As a last resort, simply terminating the worker and immediately restarting when execution is interrupted. You can see this in action by clicking 'Stop' in futurecoder as it doesn't have cross-origin isolation so `setInterruptBuffer` isn't an option. If you try to immediately run code again afterwards, there'll be a bit of a delay (unless you hit case 1 above) as it reloads, but not too much.

**Comment by hoodmane** [comment]
>  I'd be curious about the performance impact of regularly checking that throughout Python execution.

My thought would be to put it on a delay "clock" to actually perform the check only once every n times, where n is chosen so that keyboard interrupt has a resolution around ~100milliseconds (say). I think that the performance cost of this would be very low.

**Comment by redthing1** [comment]
I would love to see this officially supported.

**Comment by pmp-p** [comment]
@redthing1 i fear it can't "officially" : there are too many pitfalls, the "spookiest" being well explained here https://github.com/pyscript/pyscript/issues/324#issuecomment-1151141981

Personnally i thought using a stackless vm could solve problems I/O somehow,  but it only does solve *some* of them. There are still problems with C modules doing python callbacks or not giving hand (vsync) for rendering+I/O processing. 

Tbh i think wasm vm should be fixed instead of trying to bring a whole OS with task switching at the vm python layer or emulating spinlocks with web workers. The clean way is explained here https://lwn.net/Articles/671657/

**Comment by jimmywarting** [comment]
https://v8.dev/blog/jspi

**Comment by rth** [comment]
> [v8.dev/blog/jspi](https://v8.dev/blog/jspi)

Yes, there is work to add JS Promise integration  in  https://github.com/pyodide/pyodide/pull/3210 but it's all depending on stable browser support timeline. Something that requires users to enable an experimental flag in Chrome is not going to be immediately useful.

**Comment by jimmywarting** [comment]
Could we create a beta wasm version so folks who are willing to enable the flag can test it out?
I'm running on a clean slate here, so i'm not worried about breaking anything from existing users (as i have none right bnow) just want to explore the field and play around with it.

**Comment by hoodmane** [comment]
Well there are a still few strange CI issues on #3210. When I get around to fixing those and the PR is reviewed and all then it will get merged. It's a slightly large PR with a bunch of subtle behaviors so if the review is good this might take a while. In the meantime you can build a version of Pyodide from that branch if you want to try it out. The functionality works in chrome/node with the appropriate feature flag.

**Comment by jprendes** [comment]
If compatibility is a concern, you could always use [feature detection](https://github.com/GoogleChromeLabs/wasm-feature-detect) to load different version and do progressive enhancement.

**Comment by emeryberger** [comment]
Also quite late to this issue, but we solved this problem at the "system" level in Browsix (originally in JavaScript, ported forward to WASM; think of it as UNIX in your browser):

* https://browsix.org/
* https://arxiv.org/pdf/1611.07862
* https://www.usenix.org/conference/atc19/presentation/jangda
* https://github.com/plasma-umass/browsix

This research took place some time ago, and the github repo is probably not useful at this point, but the techniques we employed might be of interest.

**Comment by Ericwasepic127** [comment]
know what? Can you guys use my working fix? Only you must use `terminal worker` so [here's issue](https://github.com/pyscript/pyscript/issues/2471)
tell me whenever it works or not - plus here's working at https://ericwasepic127.github.io/hello.html

**Comment by ryanking13** [comment]
Well the default `input` function in pyodide is a `prompt` so we cannot use it (or at least people can already use `input` somehow). I believe people can apply different solutions in their downstream application using your fix but I don't think your solution is something we can globally upstream to Pyodide

**Comment by Ericwasepic127** [comment]
it at least copies and does what it wants
prompt will runs prompt and returns what it gave

**Comment by hoodmane** [comment]
I'm closing this as completed by JSPI. Of course, it doesn't work in Firefox or Safari, but we are not going to implement any other solutions in favor of waiting for Firefox and Safari to implement web standards.

**Comment by hoodmane** [comment]
Anyone is welcome to open a new issue if they have a specific problem.
